builder/dockerfile: ADD with best-effort xattrs

corhere opened the pull request:
- Related to #47137 

Archives being unpacked by Dockerfiles may have been created on other OSes with different conventions and semantics for xattrs, making them impossible to apply when extracting. Restore the old best-effort xattr behaviour users have come to depend on in the classic builder.

Buildkit will need to opt into best-effort xattr behaviour separately.

<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/moby/moby/blob/master/CONTRIBUTING.md

** Make sure all your commits include a signature generated with `git commit -s` **

For additional information on our contributing process, read our contributing
guide https://docs.docker.com/opensource/code/

If this is a bug fix, make sure your description includes "fixes #xxxx", or
"closes #xxxx"

Please provide the following information:
-->

**- What I did**

**- How I did it**

**- How to verify it**
TODO regression test

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**

thaJeztah commented:
Thinking of this; I wonder if we're trying too much here. While we could copy these attributes; they may not be portable when pushing the image. ISTR there even was code that explicitly excludes xattrs (with a limited set of allowed ones) for portability 🤔 (but I'd have to dig where). So perhaps we shouldn't even do so; otherwise we may end up with attributes that are available locally, but that get discarded after pushing / pulling the image.

Here's some prior tickets / comments that seem relevant related to that;

-https://github.com/moby/moby/issues/20331#issuecomment-223306730
- https://github.com/moby/moby/issues/1070
- https://github.com/moby/moby/issues/35699#issuecomment-349175742
- https://github.com/moby/moby/issues/35699#issuecomment-351218529

corhere commented:
We've been copying xattrs when `ADD`ing a tarball for the past ten years, as far as I can tell. How is reverting to the _existing_ pre-v25 behaviour "trying too much"?

tonistiigi reviewed (commented):
Doesn't this affect the pull flows as well (and load, import, build context transfer in legacy builer etc)? Does it make sense to add new `WithBestEffortXattrs` rather than just restoring the old xattr behavior.

corhere committed:
builder/dockerfile: ADD with best-effort xattrs

Archives being unpacked by Dockerfiles may have been created on other
OSes with different conventions and semantics for xattrs, making them
impossible to apply when extracting. Restore the old best-effort xattr
behaviour users have come to depend on in the classic builder.

The (archive.Archiver).UntarPath function does not allow the options
passed to Untar to be customized. It also happens to be a trivial
wrapper around the Untar function. Inline the function body and add the
option.

Signed-off-by: Cory Snider <csnider@mirantis.com>

corhere commented:
[Adding a parameter—even a variadic one—is an API breaking change](https://go.dev/blog/module-compatibility), and therefore would be unsuitable for backporting to v25. I'm going to rewrite this PR to not touch the pkg/archive API

thaJeztah reviewed (approved):
LGTM
